src/dune_engine/action.ml

```diff
@@ -292,31 +304,68 @@ module Full = struct
       ; locks = []
       ; can_go_in_shared_cache = true
       ; sandbox = Sandbox_config.default
+      ; patch_back_source_tree = false
       }
 
-    let combine { action; env; locks; can_go_in_shared_cache; sandbox } x =
+    let combine
+        { action
+        ; env
+        ; locks
+        ; can_go_in_shared_cache
+        ; sandbox
+        ; patch_back_source_tree
+        } x =
       { action = combine action x.action
       ; env = Env.extend_env env x.env
       ; locks = locks @ x.locks
       ; can_go_in_shared_cache =
           can_go_in_shared_cache && x.can_go_in_shared_cache
       ; sandbox = Sandbox_config.inter sandbox x.sandbox
+      ; patch_back_source_tree =
+          patch_back_source_tree || x.patch_back_source_tree
       }
   end
 
   include T
   include Monoid.Make (T)
 
+  let adapt_action_for_patch_back_source_tree t =
+    if not t.patch_back_source_tree then
+      t
+    else
+      (* Rules that patch back the source tree cannot go in the shared cache *)
+      let can_go_in_shared_cache = false in
+      (* Rules that patch back the source tree must be sandboxed in copy
+         mode. *)
+      let sandbox = Sandbox_mode.Set.singleton (Some Copy) in
+      { t with can_go_in_shared_cache; sandbox }
```

Comment:
Here we are throwing away the previous `sandbox` setting, and it's not obvious this is the right thing to do (perhaps, we should `inter` them?).

Reply:
intersection seems good. Done